pkg/endpoint/bpf.go

```diff
@@ -813,16 +813,11 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext) (he
 		if err != nil {
 			return false, err
 		}
-		// Clean up map contents
-		e.getLogger().Debug("flushing old PolicyMap")
-		e.policyDebug(nil, "runPreCompilationSteps flushing old PolicyMap")
-		err = e.policyMap.DeleteAll()
-		if err != nil {
-			return false, err
-		}
 
-		// Also reset the in-memory state of the realized state as the
-		// BPF map content is guaranteed to be empty right now.
+		// Reset the in-memory state of the realized state, the realized
+		// state and BPF map entries will be filled with desired policy in
+		// the following e.syncPolicyMap, and stale entries in BPF map will
+		// be removed by the subsequent syncPolicyMapController.
```

Comment:
I agree that flushing the policymap like we did previously here doesn't sound like the right approach since it can cause transient policy drops.

That said I don't think it's a good idea to rely on `syncPolicyMapController` to implement the policy correctly. It only runs once a minute, and in an ideal world it would never exist at all. We have talked about changing logic like the `syncPolicyMapController` to warn any time that it makes a change to the policy, because this implies that the core policy logic is not properly implementing the policy in the first place.

Skipping the delete and then reinitializing the in-memory view of the current contents of the BPF map (and reinitializing pressure metrics) just puts this logic out-of-sync with the actual current state in the datapath. The whole point of "realizedState" is that it is tightly bound to the current state of the active BPF map, but without having to do expensive syscall operations to perform lookups and checks.

Here's an alternative proposal: Set some flag like `e.ForcePolicyMapSync` so that when the next `e.syncPolicyMap()` runs, it will sweep the map and delete all entries that do not have corresponding entries in the `desiredPolicyMap`. We can't do this step here, because the policy didn't get calculated yet; it's calculated in `e.regeneratePolicy()` below. But the actual `e.syncPolicyMap()` occurs even later.

Reply:
Thinking a bit more about this... Another way to look at it is that upon restore, the 'realizedState' of the policymap for the endpoint is out-of-sync with the datapath, because it is encoded in process memory, and the process just started up again. So a different way to ensure that the policymap synchronization occurs correctly would be from this code to read the BPF policymap and populate the 'realizedState' with the current state of the BPF map. Then the 'realized' state will be accurate and be bound closely with the BPF implementation.

Then we don't need any additional flag, next time the `e.syncPolicyMap()` runs, it will automatically deal with the discrepancies between the 'desired' and 'realized' state.